[quant][graph] Add JIT passes for dynamic quant multi uses of quant node

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37125 [quant][graph] Add JIT passes for dynamic quant multi uses of quant node**
* #37093 [quant][graph] Run dynamic quantization for specific ops
* #37045 [graph][quant] Enable accessing child/grandchild modules in forward
* #37014 [quant][graph] Add check for qconfig_dict key

Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21283697](https://our.internmc.facebook.com/intern/diff/D21283697)

supriyar committed:
[quant][graph] Add JIT passes for dynamic quant multi uses of quant node
Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2993,17 +3059,55 @@ void FoldQuantizedPrepackingOps(Module& module) {
   PrePackingOpsFolder(module, filter_fn, "quantized");
 }
 
+void RemoveRedundantQuantizeOps(
```
you can split this to a separate PR

supriyar replied:
I feel this is still related to the change of replicating choose_qparams and quant. We shouldn't have a case of extra choose_qparams and quant nodes in the graph because it will affect perf. The test checks for this as well.
If you think splitting it helps in some other way then let me know.

jerryzh168 replied:
it's a different pass though, you can still land them together even after you split. 
But I don't have a strong preference here.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2993,17 +3059,55 @@ void FoldQuantizedPrepackingOps(Module& module) {
+        "Expect dequant output to have single use");
+    Node* user = dequant_out->uses()[0].user;
+    if (!nodeQuantizable(user, is_dynamic)) {
+      SubgraphRewriter rewriter;
+      rewriter.RegisterRewritePattern(
+          dynamic_quant_ops.pattern_string, dynamic_quant_replacement);
+      rewriter.runOnGraph(graph);
```
you can add a filter for this rewrite

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2709,6 +2716,66 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
+    Value* original_val = n->inputs()[0];
+    Value* quantized_val = n->output();
+    Value* dtype = n->inputs()[3];
+    TORCH_CHECK(original_val->uses().size() == 2);
```
what is this check?

supriyar replied:
The original value should only have 2 uses i.e. choose_qparams and quantize_per_tensor. I

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2709,6 +2716,66 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
```
I think it's better to check for the call sequence of "choose_qparams - quant - dequant" here directly

jerryzh168 replied:
you can use subgraph matcher here as well

supriyar replied:
At this point we have already called replicateDequant so the dequant nodes have been replicated. This way we can re-use that code from static and add this as a special case for dynamic

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -3011,16 +3078,48 @@ void FoldQuantizedPrepackingOps(Module& module) {
   SwapDeQuant(graph);
   InsertPrepackUnpack(graph);
   ConstantPropagation(graph);
+  RemoveRedundantQuantizeOps(graph, is_dynamic);
```
can we do this immediately after ReplicateChooseQParamsQuant?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2709,6 +2716,66 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
+    for (const Use& use : uses) {
+      auto* user = use.user;
+      WithInsertPoint ins(user);
+      Node* choose_qparam =
+          insertChooseQParamsCall(graph.get(), original_val, user);
+      std::vector<Value*> quant_inputs = {original_val};
+      for (auto& out : choose_qparam->outputs()) {
+        quant_inputs.push_back(out);
+      }
+      quant_inputs.push_back(dtype);
+      Node* quant = graph->create(n->kind(), quant_inputs);
+      quant->output()->setDebugName(original_val->debugName() + ".quant");
+      graph->insertNode(quant);
+      user->replaceInputWith(quantized_val, quant->output());
```
I think there should be similar code when you insert choose_qpram - quant - dequant in the beginning, can you refactor this into a function?

jerryzh168 replied:
probably need some refactor of insertQuantDeQuantCall

jerryzh168 reviewed (changes requested):
please match the call sequence "choose_qparam - quant - dequant' as discussed in chat.

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1834,6 +1834,26 @@ def forward(self, x):
                    .check("quantized::linear_dynamic") \
                    .run(model.graph)
 
+    def test_dynamic_quant_multi_uses(self):
```
this is a test for finalize pass, so can be considered as a unit test.

supriyar replied:
Yeah I'll refactor the dynamic quant tests in a follow-up PR.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2708,6 +2740,68 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
+    }
+  }
+  for (const auto& n : nodes_to_rewrite) {
+    // dequantize
```
nit: can you unpack `n` here to `dequantize, quantize, choose_qparams` and then call remove on these things? similar for the next loop.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2708,6 +2740,68 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
+        %a_dequant = aten::dequantize(%a_quant)
+        return (%a_dequant) )");
+  const Graph& dynamic_quant_graph = *dynamic_quant_pattern.pattern_graph;
+  const auto& vmap = dynamic_quant_pattern.vmap;
+  Value* dequant_val = vmap.at("a_dequant");
+  Node* pattern_dequant = dequant_val->node();
+  Value* quant_val = vmap.at("a_quant");
+  Node* pattern_quant = quant_val->node();
+  Value* choose_qparam_val = vmap.at("a_scale");
+  Node* pattern_choose_qparam = choose_qparam_val->node();
```
nit: these can be moved to after the check for matches

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -3010,13 +3104,46 @@ void FoldQuantizedPrepackingOps(Module& module) {
-  ReplicateDeQuant(graph);
+  if (!is_dynamic) {
+    ReplicateQuant(graph);
+    ReplicateDeQuant(graph);
```
I feel ReplicateDequant is still useful in case the weight is used by multiple ops?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -3010,13 +3104,46 @@ void FoldQuantizedPrepackingOps(Module& module) {
+    ReplicateQuant(graph);
+    ReplicateDeQuant(graph);
+  } else {
+    ReplicateChooseQParamsQuantDequant(graph);
+  }
+  RemoveRedundantQuantizeOps(graph, is_dynamic);
```
I think we just need to run these two passes before ReplicateDeQuant

supriyar replied:
Sounds good

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1550,24 +1583,23 @@ void insertQuantDeQuantCall(
+      }
     }
+    // in second pass, replace the original observed value with dequant output
+    insertDeQuantForAllUse(g, quant->output(), original_val, uses);
```
we don't need to insert dequant for all use here anymore since we have replicate dequant pass in the end.

supriyar replied:
Sounds good. I was wondering if I should remove this

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1486,16 +1500,44 @@ void insertDeQuantCall(
 
+using DynamicQuantOps = std::tuple<Node*, Node*, Node*>;
+
+DynamicQuantOps insertChooseQParamQuantDequant(
```
can you break this down further to insertChooseQParams and insertQuant?

jerryzh168 replied:
since insertQuant can be reused by another branch

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1515,32 +1557,23 @@ void insertQuantDeQuantCall(
+        g, v, dtype, at::Symbol::aten(quantize_func));
+
+    v->replaceAllUsesWith(original_val);
+    std::vector<Use> uses = original_val->uses();
+    for (const auto& use : uses) {
+      auto* user = use.user;
+      if (user != std::get<1>(quant_nodes) && user != observer &&
+          user != std::get<0>(quant_nodes)) {
+        user->replaceInputWith(
+            original_val, std::get<2>(quant_nodes)->output());
+      }
+    }
```
I think this code can be shared by two branches?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1515,32 +1557,23 @@ void insertQuantDeQuantCall(
   } else {
     quantize_func = "quantize_per_tensor";
   }
+  Value* original_val = observer->input(1);
```
could you add comments for `original_val` and `v`, it's starting to get a bit confusing

jerryzh168 reviewed (changes requested):
RemoveRedundantQuantizeOps looks good, but insertQuantDeQuant might need more refactor

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1486,16 +1500,44 @@ void insertDeQuantCall(
+  return std::make_tuple(choose_qparams, quant, dequant);
+}
+
 void insertQuantDeQuantCall(
```
also this function should probably be renamed now, maybe things like insertNumericsChangingOps

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -762,15 +762,15 @@ def forward(self, x, w0, w1, w2):
-        # we just check we have one dequant on every op input, even input
-        # is sharded as multi uses
-        FileCheck().check_count("aten::dequantize", 9, exactly=True) \
+        m = wrap_cpp_module(torch._C._jit_pass_insert_quant_dequant(m._c, "forward", False))
+        # We replicate dequant in finalize call
+        m = wrap_cpp_module(torch._C._jit_pass_quant_finalize(m._c))
+        # check that quantized ops are inserted correctly.
+        FileCheck().check_count("quantized::conv2d_prepack", 3, exactly=True) \
                    .run(str(get_forward_graph(m._c)))
```
this test can be removed actually, we already have tests for replicate_dequant

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1475,7 +1475,32 @@ void InsertObserversHelper::propagateObservedProperty(
-void insertDeQuantCall(
+Node* insertQuant(
+    Graph* graph,
+    std::vector<Value*> inputs,
```
please fix lint

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1486,17 +1511,44 @@ void insertDeQuantCall(
+    Value* original_val,
+    Value* dtype,
+    NodeKind quant_kind) {
+  std::string choose_qparams_func = "_choose_qparams_per_tensor";
+  auto reduce_range = graph->insertConstant(false);
+  // choose_qparams_per_tensor has 2 outputs, (scale, zero_point).
+  Node* choose_qparams = graph->create(
+      at::Symbol::aten(choose_qparams_func),
+      {original_val, reduce_range},
+      /* num_outputs = */ 2);
+  choose_qparams->output(0)->setDebugName(original_val->debugName() + ".scale");
+  choose_qparams->output(0)->setType(FloatType::get());
+  choose_qparams->output(1)->setDebugName(
+      original_val->debugName() + ".zero_point");
+  choose_qparams->output(1)->setType(IntType::get());
+  graph->insertNode(choose_qparams);
```
to be more consistent, could you add a function for insertChooseQParams as well?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1505,69 +1557,43 @@ void insertQuantDeQuantCall(
-        use.user != choose_qparams) {
-      uses.push_back(use);
+  observer_out->replaceAllUsesWith(original_val);
+  std::vector<Use> uses = original_val->uses();
+  for (const auto& use : uses) {
```
nit: you can merge these two lines now

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -3010,11 +3100,43 @@ void FoldQuantizedPrepackingOps(Module& module) {
   ConstantPropagation(graph);
+  if (is_dynamic) {
+    ReplicateChooseQParamsQuantDequant(graph);
+    RemoveRedundantQuantizeOps(graph, is_dynamic);
```
`is_dynamic` seems to be redundant here?

supriyar replied:
Yes it is, I was just trying to separate out the code paths. But can remove it if you prefer that.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -3010,11 +3100,43 @@ void FoldQuantizedPrepackingOps(Module& module) {
+        dequant_out->uses().size() == 1,
+        "Expect dequant output to have single use");
+    Node* user = dequant_out->uses()[0].user;
+    return !nodeQuantizable(user, is_dynamic);
```
should we just use is_dynamic=True here?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1486,17 +1511,44 @@ void insertDeQuantCall(
+  return std::make_tuple(choose_qparams, quant, dequant);
+}
+
+void insertQuantizationOps(
```
Thanks! this function looks pretty clear now

jerryzh168 reviewed (approved):
Had a few inline comments, please address them before landing.

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add JIT passes for dynamic quant multi uses of quant node"


Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
